### DISCUSSION TITLE
What is the replacement for `ComponentConfiguration.onRemove(ComponentType)`?

### DISCUSSION THREAD

**Discussion by LeHaine** [Quillraven/Fleks#162] [Q&A] [answered]
I've just updated for the first time in awhile and had some configuration that handled resetting a component and adding it back to a pool when it was removed via: `ComponentConfiguration.onRemove(Component)`. Is the replacement the lifecycle functions in the `Component<T>` `onAdd(Entity)` and `onRemove(Entity)`? If so, is there a way we can still hook into a `Component` removal from an entity from a world configuration standpoint?

**Comment by Quillraven** [comment]
I am not 100% sure what you mean. The `onAdd` and `onRemove` functions of a component run in the context of a world already.

**Reply by LeHaine** [reply]
I was doing this:

```kotlin
world {
    components {
        onAdd(AttachComponent, AttachComponent.onAdd)
        addComponentToPoolOnRemoveHooks()
    }
    // ....
}

fun ComponentConfiguration.addComponentToPoolOnRemoveHooks() {
    onRemove(SpriteComponent) { _, cmp -> spritePool.free(cmp) }
    onRemove(ProjectileComponent) { _, cmp -> projectilePool.free(cmp) }
    // ....
}
```
But now that seems to be gone.

**Reply by Quillraven** [reply, marked as answer]
Ah okay that got replaced by the component lifecycle functions. In your case you need onAdd on your attach component and onRemove on your Sprite and Projectile component.

If your pools are not globally accessible then add them as injectables to your world and get them via `inject` in the lifecycle function.
